File: Breaknes/Breaknes/SoundProcessing.cs
using SharpTools;
using System.IO;
using System.Reflection.Metadata;
using System.Windows.Forms;

// Since the simulation is now very slow, the following strategy is used to reproduce the sound:
// Collect one second of sound. As it is recorded, play it back via DirectSound and clear the receive buffer.

namespace Breaknes
{
	public class AudioRender
	{
		private DSound? audio_backend;
		private int OutputSampleRate = 48000;
		private float OutputDC = 0.0f;
		private List<float> SampleBuf = new();
		public bool Dma = false;       // atomic

		private BreaksCore.AudioSignalFeatures aux_features;
		private int DecimateEach = 1;
		private int DecimateCounter = 0;

		private bool dump_audio = false;
		private string dump_audio_dir = "";
		private string dump_rom_name;

		private SimpleIIR filter = new SimpleIIR();
		private bool iir = true;

		public AudioRender(System.IntPtr handle, bool dump, string dump_dir, string rom_name, bool use_iir, int cutoff_freq)
		{
			dump_audio = dump;
			dump_audio_dir = dump_dir;
			dump_rom_name = rom_name;

			audio_backend = new DSound(handle);

			BreaksCore.GetApuSignalFeatures(out aux_features);

			iir = use_iir;
			if (iir)
			{
				filter.Reconfigure(aux_features.SampleRate, cutoff_freq);
			}

			// SRC
			Redecimate();
		}

		/// <summary>
		/// This sampler is used for SRC.
		/// The AUX output is sampled at a high frequency, which cannot be played by a ordinary sound card.
		/// Therefore, some of the samples are skipped to match the DSound playback frequency (SourceSampleRate variable).
		/// </summary>
		public void FeedSample()
		{
			if (iir)
			{
				float sample;
				BreaksCore.SampleAudioSignal(out sample);
				filter.Input(sample);
			}

			if (DecimateCounter >= DecimateEach)
			{
				float sample;
				if (iir)
				{
					sample = filter.Output();
				}
				else
				{
					BreaksCore.SampleAudioSignal(out sample);
				}
				SampleBuf.Add(sample);
				DecimateCounter = 0;

				if (SampleBuf.Count >= OutputSampleRate)
				{
					Playback();
				}
			}
			else
			{
				DecimateCounter++;
			}
		}

		/// <summary>
		/// If you change the sampling frequency of the sound source or output frequency, you must recalculate the decimation factor.
		/// </summary>
		private void Redecimate()
		{
			BreaksCore.GetApuSignalFeatures(out aux_features);
			DecimateEach = aux_features.SampleRate / OutputSampleRate;
			Console.WriteLine("APUSim sample rate: {0}, DSound sample rate: {1}, decimate factor: {2}", aux_features.SampleRate, OutputSampleRate, DecimateEach);
			DecimateCounter = 0;
		}

		private void Playback()
		{
			if (audio_backend != null)
			{
				Console.WriteLine("Play 1 second");
				Dma = true;
				DumpAudio();
				audio_backend.PlaySampleBuf(OutputSampleRate, SampleBuf, OutputDC);
				SampleBuf.Clear();
				Dma = false;
			}
		}

		private void DumpAudio()
		{
			if (dump_audio)
			{
				int numberOfSamples = SampleBuf.Count;
				byte[] bytes = new byte[numberOfSamples * 2];

				for (int i = 0; i < numberOfSamples; i++)
				{
					short value = (short)((SampleBuf[i]) * Int16.MaxValue);

					bytes[2 * i] = (byte)((value >> 0) & 0xff);
					bytes[2 * i + 1] = (byte)((value >> 8) & 0xff);
				}

				using (var stream = new FileStream(dump_audio_dir + "/" + dump_rom_name + "_aux.bin", FileMode.Append))
				{
					stream.Write(bytes, 0, bytes.Length);
				}
			}
		}
	}
}

File: Breaknes/BreaknesSDL/SoundProcessing.cpp
#include "pch.h"

// Since the simulation is now very slow, the following strategy is used to reproduce the sound:
// Collect one second of sound. As it is recorded, play it back via sound driver and clear the receive buffer.

SoundOutput::SoundOutput()
{
	Redecimate();

	SampleBuf_Size = OutputSampleRate;
	SampleBuf = new int16_t[SampleBuf_Size];
	memset(SampleBuf, 0, SampleBuf_Size * sizeof(int16_t));
	SampleBuf_Ptr = 0;

	if (SDL_InitSubSystem(SDL_INIT_AUDIO) < 0) {
		printf ("SDL audio could not initialize! SDL_Error: %s\n", SDL_GetError());
		return;
	}

	spec.freq = OutputSampleRate;
	spec.format = AUDIO_S16;
	spec.channels = 1;
	spec.samples = SampleBuf_Size;
	spec.callback = Mixer;
	spec.userdata = this;

	dev_id = SDL_OpenAudioDevice(NULL, 0, &spec, &spec_obtainted, 0);
	SDL_PauseAudioDevice(dev_id, 1);
}

SoundOutput::~SoundOutput()
{
	SDL_CloseAudioDevice(dev_id);
	SDL_QuitSubSystem(SDL_INIT_AUDIO);
	delete[] SampleBuf;
}

void SoundOutput::Mixer(void* thisptr, Uint8* stream, int len)
{
	SoundOutput* snd_out = (SoundOutput*)thisptr;
	if (snd_out->Dma) {
		SDL_MixAudioFormat(stream, (const Uint8*)snd_out->SampleBuf, AUDIO_S16SYS, len, SDL_MIX_MAXVOLUME);
		snd_out->Dma = false;
	}
	else {
		memset(snd_out->SampleBuf, 0, snd_out->SampleBuf_Size * sizeof(int16_t));
		SDL_PauseAudioDevice(snd_out->dev_id, 1);
	}
}

void SoundOutput::FeedSample()
{
	if (DecimateCounter >= DecimateEach)
	{
		float sample;
		{
			SampleAudioSignal(&sample);
		}

		SampleBuf[SampleBuf_Ptr++] = (int16_t)(sample * (float)INT16_MAX);
		DecimateCounter = 0;

		if (SampleBuf_Ptr >= SampleBuf_Size)
		{
			Playback();
		}
	}
	else
	{
		DecimateCounter++;
	}
}

/// <summary>
/// If you change the sampling frequency of the sound source or output frequency, you must recalculate the decimation factor.
/// </summary>
void SoundOutput::Redecimate()
{
	APUSim::AudioSignalFeatures aux_features{};
	GetApuSignalFeatures(&aux_features);
	DecimateEach = aux_features.SampleRate / OutputSampleRate;
	printf ("APUSim sample rate: %d, SoundCard sample rate: %d, decimate factor: %d\n", aux_features.SampleRate, OutputSampleRate, DecimateEach);
	DecimateCounter = 0;
}

void SoundOutput::Playback()
{
	printf("Play 1 second\n");
	Dma = true;
	SDL_PauseAudioDevice(dev_id, 0);
	SampleBuf_Ptr = 0;
}
